### DISCUSSION TITLE
RFC - Support for UDP port forwarding

### DISCUSSION THREAD

**Discussion by balajiv113** [lima-vm/lima#2402] [Ideas]
**Problem**
Currently port forwarding is done via ssh. Which supports only TCP port forwarding.
Due to this Lima is limited in a way that we cannot support UDP port forwarding

**Approach**
Currently lima do already have communication layer between hostagent and guestagent.

We can use the same medium to tunnel data in general. This way we can provide support for TCP, UDP and any other protocol if needed.

Task is to listen for tcp/udp on host when the event arrives and pass it through GRPC and do the reverse in guest.

**Advantages**
- No need of a separate process to do port forwarding
- GRPC does already support bi-directional streaming
- Uses vsock in case of vz driver

**Comment by balajiv113** [comment]
I have done a initial implementation to support TCP forwarding (Still in initial stage)
On checking performance, this model performs well when compared to SSH

GRPC TCP - ~3.80 Gbits/sec
GRPC TCP Reverse - ~4.77 Gbits/sec
SSH TCP - ~3.38 Gbits/sec
SSH TCP Reverse - ~3.08 Gbits/sec

**Comment by brandonros** [comment]
How could one test this branch? I wonder if `limactl validate` could fail if it detects portForwards with `proto: udp` if it is known to not work in the current state.

edit: https://lima-vm.io/docs/config/port/ `LIMA_SSH_PORT_FORWARDER=false`
